Reorganize source code // add asset-loader // Bug fixes // Clean-up

MadByteDE opened the pull request:
## Overview
Okay, here is one big PR for you guys :scream: .
Maybe I was a bit too motivated today. :smile: 

This PR contains the asset loader and pretty much a facelift for lots and lot of files.
Since I knew I had to change a lot to get the asset-loader in and doing anything, I didn't bother creating separate commits which I regret now, even it it would have been hard to separate them.
I think it would've been much easier for you to find the differences between old and new code. But it's too late now I guess.

General discussions:
#150

I hope I managed to squash all major bugs! 


## Changes
Here is a quick rundown of the biggest changes in this PR:

- **New asset loader added!**: See section below changes (but should be pretty self-explanatory in itself)
- **New module: smoke.lua** - contains the smoke code but is basically a rewrite because I wanted to merge the way animations are created. I also tried to write it as modular as possible. But there's still some work to do for it to be fully modular.
- **New modules: base.lua / building.lua**: I didn't dedicate much time to these two. I just wanted to get rid of the drawobjects.lua and decided to move them to their own files. More work needed to make them fully modular.
- **Major main.lua clean-up**
- **_Removed drawobjects.lua_**
- Made worldOffset a global: This allowed me to get rid of quite a bit of code and made the source much more readable imo. _I know_ that adding globals is not really what we want to archieve, but in that case it made sense imo. To counter-act, I removed some other globals on my way :stuck_out_tongue_closed_eyes: 
- clean-up in hud.lua
- changed some function and variable names
- more code optimizations like merged if-else-statements, not creating new fonts every drawcall, more local variables ...
- Fixed another bug causing the player to get stuck without gameover screen
- added more debug info to the hud
- added new TODOs and FIXMEs in many files

## The asset-loader
I thought It might be helpful to write down some tips for using the asset loader. The function names and overall functionality isn't final of course.

### Assets.loadDirectory(directoryPath, filetype)
Load directory can automatically load files from a given path. Currently we have to tell the function what filetype to expect in the given directory. based on the file type, the function decides which love function to use to load the file. 

**Warning:** Currently the function doesn't check if the given filetype matches the file it sees in the directory, which can cause a crash, if e.g an image is expected but a sound file is found in the directory!

### Assets.newImage/newSound(filePath, args)
The loader uses these function internally to load files and put them in their respective table ie. Assets.images. 
But we can also use those functions to manually load files into a table to access them later from anywhere in the project.

If the love function for the file type accepts more args they can be passed to the function as well.

**btw.** if the file type is "music", the file will be loaded in "stream" mode, otherwise the mode is "static" .

### Assets.newImageSet(path, args)
This function is unique to the asset-loader. It basically creates a table containing the image, the image width and the image height. It's basically just a shortcut for:
```lua
-- Without getImageSet
local sprite = {}
sprite.image = Assets.getImage("ship")
sprite.width = spriteImage:getWidth()
sprite.height = spriteImage:getHeight()

-- With getImageSet
sprite = Assets.getImageSet("ship")
```
I'm not sure if the name  newImage**Set** is very intuitive to understand. I tried newImageData and newImageContainer but none of them sounded really good to me.

### Assets.newAnimation(name, image, width, height, frames, row, durations, onLoop)
This one is long, but it's worth it imo because otherwise the code would look like this:
```lua
-- Without
local spritesheet	= love.graphics.newImage("assets/smoke.png")
local grid		= Anim8.newGrid(10, 14, spritesheet:getWidth(), spritesheet:getHeight())
local animation 	= Anim8.newAnimation(grid('1-14', 1), 0.15)

-- With
local animation = Assets.newAnimation("smoke", 10, 14, '1-14', 1, 0.15)
```
**Note:** Assets.newAnimation does **not** save animations in a table inside the Assets module (unlike all the other new functions). This is because doing this wouldn't allow to create unique objects easily which are needed to avoid bugs regarding updating a single animation multiple times etc. 

### Assets.playSound(soundName, isLooping)
Not much to say here. Provide the fileName/soundName and tell it if it is looping or not.
The sound won't get cloned, so this needs to be done manually if you need to playback a sound multiple times at once.

### Assets.newFont(size/name, (size))
new Font is special because it takes either an integer or string first.

If the first arg is a string, the function will search for a loaded font file and return it. The second arg will be the font size in that case:
`Assets.newFont("calibri", 16)`

If the first arg is a integer, the function will create a new font object and set it's size to the given value:
```lua
Assets.newFont(6)
Assets.setFont("font6")
```


### Assets.setFont(name)
self-explanatory, but I mention it because it's unlike the other functions. 
It basically does the same as love.graphics.setFont.

The name of the font always is the combination of the fontName (name of loaded file or default "font") and it's size together:
`Assets.setFont("font32")` or `Assets.setFont("calibri16")` e.g.

togfoxy commented:
Nice. Did you not go to your real job or something?   :)

togfoxy commented:
Happy to say all the functions work with no runtime error.

Unhappy to say that this PR is causing the red terrain to flicker a lot. It's quite bad and noticeable. :(  I haven't looked into the code just yet but it seems to be the sideways scrolling. If you thrust to the left you'll see the ground flicker and when you hit the edge of the world, the scrolling stops and the flicker also stops.

This flicker isn't in v0.10.

MadByteDE commented:
> Happy to say all the functions work with no runtime error.
> 
> Unhappy to say that this PR is causing the red terrain to flicker a lot. It's quite bad and noticeable. :(  I haven't looked into the code just yet but it seems to be the sideways scrolling. If you thrust to the left you'll see the ground flicker and when you hit the edge of the world, the scrolling stops and the flicker also stops.
> 
> This flicker isn't in v0.10.

I've noticed this but wasn't sure if its something I introduced. Can't be anything difficult since I didn't change much code in functionality. Looks like a rounding issue for the x position of the lines or something. Will take a look after work.

Philbywhizz commented:
Wow.. lots to go through here... might take me a bit.

Philbywhizz commented on source/scripts/enum.lua:
```diff
@@ -0,0 +1,26 @@
+module(...,package.seeall)
```
I know this isn't your bit of code, but this line had always confused me as to what it exactly does and if it is needed

togfoxy replied:
I was waiting for someone to say something.  :)

It's old style LUA and it's how it was done "back then". I'll change it when I'm done reminiscing.  lolz

MadByteDE replied:
I'm not happy with that line either.

Its used to set an environment for the containing variables and functions and returns them as a module / table.

Same as
```lua
local enum = {}
enum.foo = bar
return enum
```

Its just to not write enum everytime.
It has a big flaw though. For it to work it has to kinda inherite the global space which means all global variables and functions can be accessed through the module like `enum.print("Foobar")` which is a problem in my book.

Philbywhizz replied:
No problem. It confused me at first because we have now have a module called modules.

Philbywhizz commented on source/main.lua:
```diff
@@ -1,155 +1,207 @@
+-- ~~~~~~~~~~~
+-- Libraries
+-- ~~~~~~~~~~~
+
```
I wonder if we should take this opportunity to make all the require statements consistant.

Some are require 'xx', some are require "xxx", some are require (" ")

togfoxy replied:
I suggest require 'xx'. Easy enough to fix in the next PR.

Philbywhizz replied:
I'm happy with '' as well - as long as it is consistant (maybe I'll add that to the code style as well)

Philbywhizz reviewed (changes requested):
This is an amazing PR. My only gripe is that it is rather large (but restructuring the code is no simple task as it touches lots of things).

I've suggested a couple of cosmetic things. Only one real bug (line 74 in menus.lua) that I came across while play testing.

This PR will definitely make it easier for us to work on bits without stepping on each other and causing merge conflicts.

togfoxy reviewed (approved):
Happy if Phil's happy.

Philbywhizz commented:
I'll be happy when the line74 menus.lua bug is fixed, then I'm happy to have this massive PR merged in. :)

MadByteDE commented:
Pushed the fixes and made some minor changes:
- Fixed flickering terrain by rounding the lander.x value before calculating gintWorldOffset
- Removed 'module(...,package.seeall)' in enum.lua 
- Reverted changed function name to Lander.create since the naming was redundant in itself anyway
- Moved assetloader.lua to lib directory
- Changed brackets for 'require' to be consistent

MadByteDE commented:
> Nice. Did you not go to your real job or something? :)

Didn't see your comment. I don't work on weekends and I hadn't much to do anyway :smile:

I think changing so many things at once on a single day will be the last time. I see some more possibilities for improvement here and there, but overall I'd say we got a pretty good looking foundation for the source code now.

togfoxy commented:
I'll give this a play test then merge it in shortly.

togfoxy commented:
No obvious problems. I can see the bit Phil reported is fixed. Will merge this into main branch. Thanks!!!
